### PR TITLE
lint: add dupSubExpr checker

### LINKED ISSUE #395: lint: add duplicated sub-expr checker
Detects suspicious duplicated sub-expressions.

**Before:**
```go
sort.Slice(xs, func(i, j int) bool {
	return xs[i].v < xs[i].v // Same indexes
})
```
**After:**
```go
sort.Slice(xs, func(i, j int) bool {
	return xs[i].v < xs[j].v // Different indexes
})
```

### PR DESCRIPTION
**Pull request by Quasilyte**
Fixes #395

### CODE REVIEW

File: lint/dupSubExpr_checker.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	if c.isSafe(expr) && c.opSet[expr.Op] && astequal.Expr(expr.X, expr.Y) {
+		c.warn(expr)
+	}
+}
+
+func (c *dupSubExprChecker) resultIsFloat(expr ast.Expr) bool {
+	if typ, ok := c.ctx.typesInfo.TypeOf(expr).(*types.Basic); ok {
+		return typ.Info()&types.IsFloat != 0
+	}
+	return false
```
Comment (cristaloleg) [inline lint/dupSubExpr_checker.go]: Can be:
```
typ, ok := c.ctx.typesInfo.TypeOf(expr).(*types.Basic)
return ok && typ.Info()&types.IsFloat != 0
```

File: lint/dupSubExpr_checker.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	// Can be extended as needed.
+	//
+	// Note that it is not very strict "safe" as
+	// index expressions are permitted even though they
+	// may cause panics.
+	switch expr := expr.(type) {
+	case *ast.BinaryExpr:
+		return c.isSafe(expr.X) && c.isSafe(expr.Y)
+	case *ast.UnaryExpr:
+		return c.isSafe(expr.X) && expr.Op != token.ARROW
```
Comment (cristaloleg) [inline lint/dupSubExpr_checker.go]: I think it's better to reorder them.
I mean `return expr.Op != token.ARROW && c.isSafe(expr.X)`

File: lint/dupSubExpr_checker.go (hunk 1 above)
Comment (Quasilyte) [reply lint/dupSubExpr_checker.go]: Thanks!
Not the first time such simplification pops up on the review.
Maybe we can add a checker for it?

File: lint/dupSubExpr_checker.go (hunk 2 above)
Comment (Quasilyte) [reply lint/dupSubExpr_checker.go]: Makes sense.

**Comment by Quasilyte** [comment]
PTAL.

**Pull request merged**